### PR TITLE
[WIP] Multi round nbest rescoring

### PR DESCRIPTION
**Pull request by pkufool**
There are several decoding methods in icefall now,  their pipelines are demonstrated  as the picture below. What I want to do in this pull request locates in the **red rectangle**, it was proposed here(https://github.com/k2-fsa/snowfall/issues/232) several weeks ago.   

image

This is just at the very beginning, just copy the related previous work from k2 & snowfall.

### CODE REVIEW

**Comment by pkufool** [comment]
After fixing some errors, now I can get the same WER using nbest attention rescoring, see the table below:

|Decoding method|WER(%) test-clean|WER(%) test-other|Notes|
|--|--|--|--|
|rescore with attention decoder (num-paths=500, max-duration=1, lattice-score-scale=0.5)|2.57|5.94|the original attention rescoring method|
|rescore all nbest path with attention decoder(num-paths=500, max-duration=1, lattice-score-scale=0.5)|2.57|5.98|rescore all the unique nbest path, should be the same with the above row, just to check the function `rescore_nbest_with_attention_rescorer` is right|
|rescore topk * 2 nbest  path with attention decoder(topk=10,num-paths=500, max-duration=1, lattice-score-scale=0.5)|2.59|5.97|rescore the top2k nbest path (sorted by lattice scores) with attention decoder|
|rescore topk + topk nbest path with attention decoder(topk=10, num-paths=500, max-duration=1, lattice-score-scale=0.5)|2.59|6.03|rescore topk nbest path (sorted by lattice scores) first, select the other topk path by a small estimation model, then rescore the selected topk nbest path with attention decoder|

The future plans are:
1. Tune the small model that predicting the rescoring score of a path (to select better path).
2. Support batch processing (only support batch-size=1 now).
3. Check the code to speed up the decoding (now it spends more time than original attention-decoder method, 29mins vs 20mins or so).

File: icefall/score_estimator.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    str2bool,
+)
+
+
+class Dataset(torch.utils.data.Dataset):
+    def __init__(
+        self,
+        path: Path,
+        model: str,
+    ) -> None:
```
Comment (csukuangfj) [inline icefall/score_estimator.py:23]: Could you add documentations to the methods/functions in this file?

File: icefall/score_estimator.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
+import argparse
+import glob
+import logging
+from pathlib import Path
+from typing import Tuple, List
```
Comment (csukuangfj) [inline icefall/score_estimator.py:5]: Could you have a look at
https://icefall.readthedocs.io/en/latest/contributing/code-style.html
to follow the code style?

File: icefall/score_estimator.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        path: Path,
+        model: str,
+    ) -> None:
+        super().__init__()
+        files = sorted(glob.glob(f"{path}/*.pt"))
+        if model == 'train':
+            self.files = files[0: int(len(files) * 0.8)]
+        elif model == 'dev':
+            self.files = files[int(len(files) * 0.8): int(len(files) * 0.9)]
+        elif mode == 'test':
```
Comment (csukuangfj) [inline icefall/score_estimator.py:30]: `mode` or `model`?

File: icefall/score_estimator.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            in_features=hidden_dim,
+            out_features=2
+        )
+        self.sigmod = nn.Sigmoid()
+
+    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        x = self.embedding(x)
+        x = self.sigmod(x)
+        x = self.output(x)
+        mean, var = x[:, 0], x[:, 1]
```
Comment (csukuangfj) [inline icefall/score_estimator.py:67]: `var` -> `log_var`?

File: icefall/score_estimator.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+        type=int,
+        default=5,
+        help="Dim of input feature.",
+    )
+
+    parser.add_argument(
+        "--hidden-dim",
+        type=int,
+        default=20,
+        help="Neural number of didden layer.",
```
Comment (csukuangfj) [inline icefall/score_estimator.py:88]:
```suggestion
        help="Neural number of hidden layer.",
```

File: icefall/score_estimator.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+    torch.cuda.manual_seed(42)
+
+    setup_logger(f"{args.exp_dir}/rescore/log")
+
+    model = ScoreEstimator(
+        input_dim = args.input_dim,
+        hidden_dim = args.hidden_dim
+    )
+
+    model = model.to("cuda")
```
Comment (csukuangfj) [inline icefall/score_estimator.py:134]: Shall we support also CPU?

File: icefall/score_estimator.py [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+            optimizer.step()
+            training_loss += loss.item()
+            step += len(y)
+        training_loss /= step
+
+        dev_loss = 0.0
+        step = 0
+        model.eval()
+        for x, y in dev_dataloader:
+            mean, var = model(x.cuda())
```
Comment (csukuangfj) [inline icefall/score_estimator.py:169]: I would recommend you to define a variable `device`, which can be either a CPU or a CUDA device, and use `model(x.to(device))`.

File: icefall/score_estimator.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+            loss.backward()
+            optimizer.step()
+            training_loss += loss.item()
+            step += len(y)
+        training_loss /= step
+
+        dev_loss = 0.0
+        step = 0
+        model.eval()
+        for x, y in dev_dataloader:
```
Comment (csukuangfj) [inline icefall/score_estimator.py:168]: Please put the evaluation process in a context disabling gradient computation.

File: icefall/score_estimator.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
+import argparse
+import glob
+import logging
+from pathlib import Path
+from typing import Tuple, List
+
+import torch
+import torch.nn as nn
+from torch.utils.data import DataLoader
+from torch.utils.tensorboard import SummaryWriter
```
Comment (csukuangfj) [inline icefall/score_estimator.py:10]: Please remove unused imports.
If you have a look at https://icefall.readthedocs.io/en/latest/contributing/code-style.html
and start to use `flake8`, it will tell you this import is never used.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: icefall/score_estimator.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+        help="Directory to store experiment data.",
+    )
+    return parser
+
+
+def main():
+    parser = get_parser()
+    args = parser.parse_args()
+    torch.manual_seed(42)
+    torch.cuda.manual_seed(42)
```
Comment (csukuangfj) [inline icefall/score_estimator.py:125]:
```suggestion
```

`torch.manual_seed(42)` already did it.
